File: precurrment vedio/procurement_agent/frontend/js/projects.js

// --- Project & Client Management ---

let currentLifecycleProjectId = null;

async function loadProjects() {
    try {
        const response = await fetch('/api/projects');
        const projects = await response.json();
        const tbody = document.getElementById('projects-table-body');
        if (!tbody) return;

        if (!projects.length) {
            tbody.innerHTML = '<tr><td colspan="5" style="text-align:center; padding:30px; color:#94a3b8;">No projects found. Fetch emails to auto-create projects from client requests.</td></tr>';
            return;
        }

        tbody.innerHTML = projects.map(p => `
            <tr>
                <td><span class="badge badge-neutral">${p.client_formatted_id}</span></td>
                <td><strong>${p.client_name}</strong></td>
                <td>
                    <strong>${p.name}</strong>
                    ${p.source_email_id ? `
                        <a href="#" onclick="viewEmailDetails(${p.source_email_id}); return false;" title="View Source Request" style="margin-left:8px; color:#60a5fa; font-size:0.8rem;">
                            <i class="fas fa-external-link-alt"></i> Original
                        </a>
                    ` : ''}
                </td>
                <td><span style="font-size: 0.85rem; color: var(--text-muted);">${p.client_email}</span></td>
                <td><span class="status-badge ${p.status}">${p.status.toUpperCase()}</span></td>
                <td>${new Date(p.created_at).toLocaleDateString()}</td>
                <td>
                    <button class="btn btn-sm btn-outline" onclick="viewProject(${p.id})">
                        <i class="fas fa-eye"></i> View
                    </button>
                </td>
            </tr>
        `).join('');
    } catch (err) {
        console.error("Load Projects Error:", err);
    }
}

// ... (loadClients removed from view but keeping for internal use if needed)

// ── View a specific project: open lifecycle modal ──
async function viewProject(projectId) {
    currentLifecycleProjectId = projectId;
    const modal = document.getElementById('lifecycle-modal');
    if (!modal) return;

    // Show modal with loading state
    modal.style.display = 'flex';
    document.getElementById('lc-project-name').textContent = 'Loading...';
    document.getElementById('lc-emails-list').innerHTML = '<p style="color:#94a3b8; font-size:0.85rem;">Loading...</p>';
    document.getElementById('lc-drafts-list').innerHTML = '<p style="color:#94a3b8; font-size:0.85rem;">Loading...</p>';
    document.getElementById('lc-responses-list').innerHTML = '<p style="color:#94a3b8; font-size:0.85rem;">Loading...</p>';

    try {
        const res = await fetch(`/api/projects/${projectId}/lifecycle`);
        if (!res.ok) throw new Error('Failed to load lifecycle');
        const data = await res.json();
        renderLifecycleModal(data);
    } catch (err) {
        console.error('viewProject error:', err);
        showToast('Failed to load project details.', 'error');
        modal.style.display = 'none';
    }
}

// ── Render lifecycle data into the modal ──
function renderLifecycleModal(data) {
    // Project header
    document.getElementById('lc-project-name').textContent = data.project.name;
    document.getElementById('lc-project-meta').textContent = `Created ${new Date(data.project.created_at).toLocaleDateString()} · Project ID: ${data.project.id}`;
    document.getElementById('lc-project-status').textContent = data.project.status.toUpperCase();
    document.getElementById('lc-project-status').className = `lc-status-badge lc-status-${data.project.status}`;

    // Client info
    document.getElementById('lc-client-name').textContent = `${data.client.name} (ID: ${data.client.id})`;
    document.getElementById('lc-client-email').textContent = data.client.email || '';

    // De-duplicate emails (by subject and body snippet) for clean UI
    const uniqueEmails = [];
    const seenEmails = new Set();
    data.emails.forEach(e => {
        const key = `${e.subject}|${e.body.substring(0, 100)}`;
        if (!seenEmails.has(key)) {
            seenEmails.add(key);
            uniqueEmails.push(e);
        }
    });

    // Counts
    document.getElementById('lc-email-count').textContent = uniqueEmails.length;
    document.getElementById('lc-draft-count').textContent = data.drafts.length;
    document.getElementById('lc-response-count').textContent = data.vendor_responses.length;
    document.getElementById('lc-quote-count').textContent = data.comparison_data.length;

    // Emails
    const emailList = document.getElementById('lc-emails-list');
    emailList.innerHTML = uniqueEmails.length
        ? uniqueEmails.map(e => `
            <div class="lc-item">
                <strong><i class="fas fa-envelope" style="color:#60a5fa;"></i> ${e.subject}</strong>
                <span class="lc-item-meta">${e.sender} · ${new Date(e.received_at).toLocaleDateString()}</span>
                <p class="lc-item-body">${e.body}</p>
            </div>`).join('')
        : '<p class="lc-empty-text">No client emails linked yet.</p>';

    // Drafts
    const draftList = document.getElementById('lc-drafts-list');
    draftList.innerHTML = data.drafts.length
        ? data.drafts.map(d => `
            <div class="lc-item">
                <strong><i class="fas fa-paper-plane" style="color:#a78bfa;"></i> ${d.subject}</strong>
                <span class="lc-item-meta">To: ${d.vendor_name} · <span class="lc-status-mini ${d.status}">${d.status.toUpperCase()}</span> ${d.sent_at ? '· Sent ' + new Date(d.sent_at).toLocaleDateString() : ''}</span>
                <p class="lc-item-body">${d.body}</p>
            </div>`).join('')
        : '<p class="lc-empty-text">No drafts generated yet.</p>';

    // Vendor Responses
    const responseList = document.getElementById('lc-responses-list');
    responseList.innerHTML = data.vendor_responses.length
        ? data.vendor_responses.map(r => `
            <div class="lc-item lc-item-green">
                <strong><i class="fas fa-reply" style="color:#34d399;"></i> ${r.subject}</strong>
                <span class="lc-item-meta">From: ${r.vendor_email} · ${new Date(r.received_at).toLocaleDateString()}</span>
                <p class="lc-item-body">${r.body}</p>
            </div>`).join('')
        : '<p class="lc-empty-text">No vendor responses yet.</p>';

    // Comparison Table
    renderLifecycleComparison(data.comparison_data);
}

function renderLifecycleComparison(quotes) {
    const table = document.getElementById('lc-comparison-table');
    const noQuotes = document.getElementById('lc-no-quotes');
    const tbody = document.getElementById('lc-comparison-body');

    if (!quotes.length) {
        table.style.display = 'none';
        noQuotes.style.display = 'block';
        return;
    }

    table.style.display = 'table';
    noQuotes.style.display = 'none';

    // Find best (lowest) price per product
    const priceByProduct = {};
    quotes.forEach(q => {
        if (!priceByProduct[q.product] || q.price < priceByProduct[q.product]) {
            priceByProduct[q.product] = q.price;
        }
    });

    tbody.innerHTML = quotes.map(q => {
        const isBest = q.price === priceByProduct[q.product];
        return `<tr class="${isBest ? 'lc-best-row' : ''}">
            <td>${q.vendor_email || '—'}</td>
            <td>${q.product}</td>
            <td><strong>${Number(q.price).toLocaleString()}</strong> ${isBest ? '<span class="lc-best-badge">BEST</span>' : ''}</td>
            <td>${q.unit || '—'}</td>
            <td style="font-size:0.8rem; color:#94a3b8;">${q.vendor_notes || '—'}</td>
        </tr>`;
    }).join('');
}

// ── View Client: show their projects ──
async function viewClient(clientId) {
    try {
        const res = await fetch(`/api/clients/${clientId}/projects`);
        if (!res.ok) throw new Error('Failed to load client projects');
        const data = await res.json();
        showToast(`${data.client.name}: ${data.projects.length} project(s). Opening first project...`, 'info');
        if (data.projects.length > 0) {
            viewProject(data.projects[0].id);
        }
    } catch (err) {
        showToast('Could not load client details.', 'error');
    }
}

// ── Close lifecycle modal ──
function closeLifecycleModal() {
    const modal = document.getElementById('lifecycle-modal');
    if (modal) modal.style.display = 'none';
    currentLifecycleProjectId = null;
}

// ── Toggle the vendor response input box ──
function toggleProcessBox() {
    const box = document.getElementById('lc-process-box');
    if (!box) return;
    box.style.display = box.style.display === 'none' ? 'block' : 'none';
}

// ── Submit vendor response for AI extraction ──
async function submitVendorResponse() {
    if (!currentLifecycleProjectId) return;

    const body = document.getElementById('lc-vendor-body').value.trim();
    const email = document.getElementById('lc-vendor-email-input').value.trim();
    const subject = document.getElementById('lc-vendor-subject').value.trim();

    if (!body || !email) {
        showToast('Please fill in vendor email and response text.', 'warning');
        return;
    }

    try {
        showToast('Extracting prices using AI...', 'info');
        const res = await fetch(`/api/projects/${currentLifecycleProjectId}/process-vendor-response`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ vendor_email: email, subject: subject || 'Vendor Quote', body })
        });

        if (!res.ok) throw new Error('Processing failed');
        const result = await res.json();

        showToast(`Success! ${result.extracted_quotes.length} price items extracted.`, 'success');

        // Clear form and refresh modal
        document.getElementById('lc-vendor-body').value = '';
        document.getElementById('lc-vendor-email-input').value = '';
        document.getElementById('lc-vendor-subject').value = '';
        document.getElementById('lc-process-box').style.display = 'none';

        // Reload the lifecycle view to show new data
        await viewProject(currentLifecycleProjectId);
    } catch (err) {
        showToast('Failed to process vendor response.', 'error');
    }
}

// --- Manual Client & Project Creation ---

function showAddClientModal() {
    const modal = document.getElementById('manual-client-modal');
    if (modal) modal.style.display = 'flex';
}

async function showAddProjectModal() {
    const modal = document.getElementById('manual-project-modal');
    if (!modal) return;
    modal.style.display = 'flex';

    // Populate client dropdown
    try {
        const response = await fetch('/api/clients');
        const clients = await response.json();
        const select = document.getElementById('mp-client-id');
        if (!select) return;

        if (!clients || !clients.length) {
            select.innerHTML = '<option value="">No clients found. Add a client first.</option>';
            return;
        }

        select.innerHTML = clients.map(c => `
            <option value="${c.id}">${c.name || 'Unknown'} (${c.email})</option>
        `).join('');
    } catch (err) {
        console.error("Failed to load clients for project modal:", err);
    }
}

async function saveManualClient() {
    const name = document.getElementById('mc-name').value.trim();
    const email = document.getElementById('mc-email').value.trim();

    if (!name || !email) {
        showToast('Please provide both name and email.', 'warning');
        return;
    }

    try {
        const response = await fetch('/api/clients', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name, email })
        });

        if (!response.ok) throw new Error('Failed to save client');

        showToast('Client added successfully!', 'success');
        hideModal('manual-client-modal');
        loadClients(); // Refresh list

        // Clear fields
        document.getElementById('mc-name').value = '';
        document.getElementById('mc-email').value = '';
    } catch (err) {
        showToast('Error saving client.', 'error');
    }
}

async function saveManualProject() {
    const name = document.getElementById('mp-name').value.trim();
    const clientId = document.getElementById('mp-client-id').value;

    if (!name || !clientId) {
        showToast('Please provide project name and select a client.', 'warning');
        return;
    }

    try {
        const response = await fetch('/api/projects', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name, client_id: parseInt(clientId) })
        });

        if (!response.ok) throw new Error('Failed to save project');

        showToast('Project created successfully!', 'success');
        hideModal('manual-project-modal');
        loadProjects(); // Refresh list

        // Clear fields
        document.getElementById('mp-name').value = '';
    } catch (err) {
        showToast('Error saving project.', 'error');
    }
}

function hideModal(modalId) {
    const modal = document.getElementById(modalId);
    if (modal) modal.style.display = 'none';
}

File: precurrment vedio/procurement_agent/backend/api/projects.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from ..database.session import get_db
from ..database.models import Project, ProjectVersion
from pydantic import BaseModel
from typing import List, Optional
from ..database import crud
from ..services.llm_client import LLMClient

router = APIRouter(tags=["Projects"])
llm_client = LLMClient()

class ProjectCreate(BaseModel):
    name: str
    client_id: int

class VendorResponseRequest(BaseModel):
    vendor_email: str
    subject: str
    body: str

@router.get("")
def list_projects(db: Session = Depends(get_db)):
    projects = db.query(Project).all()
    results = []
    for p in projects:
        # Find the first email (source) for this project
        from ..database.models import EmailRecord
        source_email = db.query(EmailRecord).join(ProjectVersion).filter(ProjectVersion.project_id == p.id).order_by(EmailRecord.id.asc()).first()
        
        results.append({
            "id": p.id,
            "name": p.name,
            "status": p.status,
            "created_at": p.created_at.isoformat() if p.created_at else None,
            "client_id": p.client_id,
            "client_formatted_id": f"CLIENT-{str(p.client_id).zfill(4)}" if p.client_id else "N/A",
            "client_name": p.client.name if p.client else "Unknown",
            "client_email": p.client.email if p.client else "N/A",
            "source_email_id": source_email.id if source_email else None
        })
    return results

@router.post("")
def create_project(req: ProjectCreate, db: Session = Depends(get_db)):
    p = crud.create_project(db, name=req.name, client_id=req.client_id)
    return {"id": p.id, "name": p.name}

@router.get("/{project_id}")
def get_project(project_id: int, db: Session = Depends(get_db)):
    p = db.query(Project).filter(Project.id == project_id).first()
    if not p: return None
    return {
        "id": p.id,
        "name": p.name,
        "status": p.status,
        "client_id": p.client_id
    }

@router.post("/{project_id}/smart-drafts")
def generate_project_smart_drafts(project_id: int, db: Session = Depends(get_db)):
    from ..services.procurement_logic import ProcurementLogic
    print(f"DEBUG: [Router] Triggering smart drafts for project {project_id}")
    
    # Get latest version
    version = db.query(ProjectVersion).filter(ProjectVersion.project_id == project_id).order_by(ProjectVersion.version_number.desc()).first()
    if not version:
        raise HTTPException(status_code=404, detail="No version found for project")
        
    logic = ProcurementLogic()
    drafts = logic.generate_smart_drafts_for_version(db, version.id)
    return {"count": len(drafts), "status": "success"}

@router.get("/{project_id}/lifecycle")
def get_project_lifecycle(project_id: int, db: Session = Depends(get_db)):
    data = crud.get_project_lifecycle(db, project_id)
    if not data:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Manual serialization to avoid 500 circular ref errors
    return {
        "project": {
            "id": data["project"].id,
            "name": data["project"].name,
            "status": data["project"].status,
            "created_at": data["project"].created_at.isoformat() if data["project"].created_at else None
        },
        "client": {
            "id": data["client"].id if data["client"] else None,
            "name": data["client"].name if data["client"] else "Unknown",
            "email": data["client"].email if data["client"] else ""
        },
        "emails": [{
            "id": e.id,
            "subject": e.subject,
            "sender": e.sender,
            "body": e.body,
            "classification": e.classification,
            "received_at": e.received_at.isoformat() if e.received_at else None
        } for e in data["emails"]],
        "drafts": [{
            "id": d.id,
            "subject": d.subject,
            "vendor_name": d.vendor_name,
            "status": d.status,
            "body": d.body,
            "sent_at": d.sent_at.isoformat() if d.sent_at else None
        } for d in data["drafts"]],
        "vendor_responses": [{
            "id": e.id,
            "subject": e.subject,
            "sender": e.sender,
            "body": e.body,
            "received_at": e.received_at.isoformat() if e.received_at else None
        } for e in data["vendor_responses"]],
        "comparison_data": data["comparison_data"] 
    }

@router.post("/{project_id}/process-vendor-response")
async def process_vendor_response(project_id: int, req: VendorResponseRequest, db: Session = Depends(get_db)):
    # 1. Extract quotes using AI
    print(f"DEBUG: [Projects] Extracting items for project {project_id} from {req.vendor_email}")
    extracted_items = llm_client.extract_quote_from_text(req.body)
    
    # 2. Get latest project version
    version = db.query(ProjectVersion).filter(ProjectVersion.project_id == project_id).order_by(ProjectVersion.version_number.desc()).first()
    if not version:
        raise HTTPException(status_code=404, detail="No project version found")

    # 3. Save extracted quotes
    saved_quotes = []
    for item in extracted_items:
        quote_data = {
            "project_version_id": version.id,
            "vendor_email": req.vendor_email,
            "product": item.get("product"),
            "price": item.get("price"),
            "unit": item.get("unit"),
            "vendor_notes": item.get("vendor_notes")
        }
        saved_quotes.append(crud.create_quoted_price(db, quote_data))
    
    return {"status": "success", "extracted_quotes": saved_quotes}
